src/shared/math.ts

```diff
@@ -0,0 +1,135 @@
+
+const six: bigint = powmod(6, 2401n, 7n)
+// n to the p is congruent to n mod a prime p,
+// so a to any power of p is as well.
+
+const fortysixish: number = natlog(100000000000000000000n)
+```
+
+### Detailed function reference
+**/
+
+import isqrt from 'bigint-isqrt'
+import {modPow} from 'bigint-mod-arith'
+
+const maxSafeNumber = BigInt(Number.MAX_SAFE_INTEGER)
+const minSafeNumber = BigInt(Number.MIN_SAFE_INTEGER)
+
+/** md
+#### safeNumber(n: number | bigint): number
+
+Returns the `number` mathematically equal to _n_ if there is one, or
+throws an error otherwise.
+**/
+export function safeNumber(n: number | bigint): number {
+    const bn = BigInt(n)
+    if (bn < minSafeNumber || bn > maxSafeNumber) {
+        throw new RangeError(`Attempt to use ${bn} as a number`)
+    }
+    return Number(bn)
+}
+
+/** md
+#### floorSqrt(n: number | bigint): bigint
+
+Returns the largest bigint _r_ such that the square of _r_ is less than or
+equal to _n_, if there is one; otherwise returns the bigint mathematically
+equal to _n_. (Thus, it leaves negative bigints unchanged.)
+**/
+export function floorSqrt(n: number | bigint): bigint {
+    const bn = BigInt(n)
+    return isqrt(bn)
+}
+
+/** md
+#### modulo(n: number | bigint, modulus: number | bigint): bigint
+
+Returns the smallest nonnegative bigint congruent to _n_ modulo
+_modulus_; this is also the remainder upon dividing _n_ by _modulus_. Note
+that this is the "mathematician's modulus" that never returns a negative
+value, unlike the built-in JavaScript `%` operator.
+
+Throws a RangeError if _modulus_ is nonpositive.
+**/
+export function modulo(n: number | bigint, modulus: number | bigint): bigint {
+    const bn = BigInt(n)
+    const bmodulus = BigInt(modulus)
+    if (bmodulus <= 0n) {
+        throw new RangeError(`Attempt to use nonpositive modulus ${bmodulus}`)
+    }
+    if (bn < 0n) {
+        return (bn % bmodulus) + bmodulus
+    }
+    return bn % bmodulus
+}
+
+/** md
+#### powmod(n, exponent, modulus): bigint
+
+All parameters have type `number | bigint`. If _modulus_ is nonpositive,
+throws a RangeError.
+
+Otherwise, if _exponent_ is positive, returns the smallest nonnegative bigint
+congruent to _n_ raised to the _exponent_ power modulo _modulus_ (but far more
+efficiently than computing `modulo(n**exponent, modulus)`).
+
+If _exponent_ is negative, first computes `i = powmod(n, -exponent, modulus)`.
+If _i_ has a multiplicative inverse modulo _modulus_, returns that inverse,
+otherwise throws a RangeError.
+**/
+export const powmod = modPow // just need to fix the name
+
+const nlg16 = Math.log(16)
+
+/** md
+#### natlog(n: number | bigint): number
+
+Returns the natural log of the input.
+**/
+export function natlog(n: number | bigint): number {
+    if (typeof n === 'number') return Math.log(n)
+    if (n < 0) return NaN
+
+    const s = n.toString(16)
+    const s15 = s.substring(0, 15)
+
+    return nlg16 * (s.length - s15.length) + Math.log(Number('0x' + s15))
+}
```

Comment:
Also, I was worrying about the algorithm I recommended and you are using here (modulo this error above).  If Math.log gives only a certain precision, what's the precision guarantee on something like this?  I'm worried it will actually have mostly incorrect digits.  In particular, if any one return from Math.log has only 16 digits of precision, how can some hyper-super-gigantic log that has 100 digits before the decimal have all 100 digits being correct?  It's a result of adding a bunch of limited-precision numbers that all start at potentially different tens-places...  (Sorry if this doesn't make any sense, we can chat in person if needed.)  Unless there's a better algorithm, I think we need to return only the 16 most significant digits -- or possibly less, because rounding errors add up and affect more significant digits eventually -- in any case, return only as many digits as we guarantee accuracy for.  Or if javascript Number is not amenable to that (e.g. how many digits of accuracy does the answer "100" imply, 3 or 1?) we should maybe also return explicitly the number of digits of accuracy the particular answer comes with in each case.

Reply:
The approximation here is that a bigint is its first 15 digits times 16^(num of digits - 15). Since Javascript number has 53 bits of precision and 53 < 15*4, this provides a full "number's worth" of precision in the answer. There is no iterating over blocks, that would be a waste of time. I am confused about your examples, I will add them to the tests and fix them if need be.

Reply:
Put all of your examples in the tests and they are all fine and full precision (as verified by Python, which also has arbitrary-precision logs).

Reply:
Yeah it was user error, I went and checked and it was my fault.  Sorry!